quartodoc/__main__.py

```diff
@@ -21,24 +22,83 @@ def get_package_path(package_name):
 
-class FileChangeHandler(FileSystemEventHandler):
+
+class FileInfo(BaseModel):
+    size: int
+    mtime: float
+    name: str= ""
+
+class QuartoDocFileChangeHandler(PatternMatchingEventHandler):
     """
     A handler for file changes.
     """
+
+    # Ignore patterns for the file watcher that are not relevant to the docs
+    py_ignore_patterns = [
+        '*/__pycache__/*',  # These are the compiled python code files which are automatically generated by Python
+        '*/.ipynb_checkpoints/*',  # This directory is created by Jupyter Notebook for auto-saving notebooks
+        '*/.vscode/*',  # If you're using Visual Studio Code, it creates this directory to store settings specific to that project.
+        '*/.idea/*',  # Similar to .vscode/, but for JetBrains IDEs like PyCharm.
+        '*/.git/*',  # i This directory is created by Git. It's not relevant to the docs.
+        '*/venv/*', '*/env/*',  '*/.env/*',  # Common names for directories containing a Python virtual environment.
+        '*/.pytest_cache/*',  # This directory is created when you run Pytest.
+        '*/.eggs/*', '*/dist/*', '*/build/*', '*/*.egg-info/*', # These are typically created when building Python packages with setuptools.
+        '*.pyo',  # These are optimized .pyc files, created when Python is run with the -O flag.
+        '*.pyd',  # This is the equivalent of a .pyc file, but for C extensions on Windows.
+        '*/.mypy_cache/*', # This directory is created when you run mypy.
+    ]
+
     def __init__(self, callback):
+        super().__init__(ignore_patterns=self.py_ignore_patterns, ignore_directories=True)
         self.callback = callback
+        self.old_file_info = FileInfo(size=-1, mtime=-1, name="")
+
+
+    def get_file_info(self, path:str) -> FileInfo:
+        """
+        Get the file size and modification time.
+        """
+        return FileInfo(size=os.stat(path).st_size, 
+                        mtime=os.stat(path).st_mtime, 
+                        name=path)
     
+    def is_diff(self, old:FileInfo, new:FileInfo) -> bool:
+        """
+        Check if a file has changed. Prevents duplicate events from being triggered.
+        """
+        same_nm = old.name == new.name
+        diff_sz = old.size != new.size
+        diff_tm = (new.mtime - old.mtime) # wait 1/4 second before triggering
+
+        if diff_tm < .25: # if consequetive events are less than 1/4th of a second apart, ignore
+            return False
+        elif same_nm:
+            if diff_sz or diff_tm >= 0.25:
+                return True
+            else:
+                return False
+        else:
+            return True     
+    
+    def callback_if_diff(self, event):
+        """
+        Call the callback if the file has changed.
+        """
+        new_file_info = self.get_file_info(event.src_path)
+        if self.is_diff(self.old_file_info, new_file_info):
+            self.callback()
+            self.print_event(event)
```

Comment:
I was thrown for a second seeing the Rebuilding message, after the build output. I wonder if putting it before might help people connect the triggering event with the build? WDYT?

For long builds, a message after might look like nothing's happening (could also sandwich the build with a pre "going to build" message, and post "just built" message)

Here is what my output looked like...

```
Watching /Users/machow/repos/quartodoc/quartodoc for changes...
2023-08-03 10:05:56,694 - quartodoc.autosummary - INFO - Generating blueprint.
2023-08-03 10:05:56,701 - quartodoc.builder.blueprint - INFO - Getting object for quartodoc:Auto
...
2023-08-03 10:05:56,798 - quartodoc.autosummary - INFO - Creating inventory file
2023-08-03 10:05:56,798 - quartodoc.autosummary - INFO - Creating inventory
2023-08-03 10:05:56,800 - quartodoc.autosummary - INFO - Writing sidebar yaml to api/_sidebar.yml
Rebuilding docs.  Detected: created path : /Users/machow/repos/quartodoc/quartodoc/a.py
```

Reply:
@machow I fixed this in https://github.com/machow/quartodoc/pull/228/commits/75935710912669ccccdedbfd61dad50b61dcbe6f

Thanks